### ISSUE TITLE
[API Proposal] add ReadOnlySpan<byte> overloads to BitConverter.ToString() 

### ISSUE THREAD

**Issue by Symbai** [dotnet/runtime#50588]
## Background and Motivation

BitConverter.ToString() has no overloads to accept Readonlyspan. When we got a span we have to call ToArray() first. Other BitConverter functions do have a Readonlyspan overload, such as BitConverter.ToInt32().

I'm not sure why ToString() was left out when adding Readonlyspan overloads to all others. At least it cannot be the question of which Encoder since ToString(byte[] value) to not provide a choice of encoder either so a Readonlyspan overload can take the same encoder ToString(byte[] value) uses.

## Proposed API

```diff
namespace System
{
    public class BitConverter {
+        public static string ToString(ReadOnlySpan<byte> value);
+        public static string ToString(ReadOnlySpan<byte> value, int startIndex);
+        public static string ToString(ReadOnlySpan<byte> value, int startIndex, int length);
     }
```

## Usage Examples

``` C#
ReadOnlySpan<byte> span = BitConverter.GetBytes("hello world").AsSpan();
Console.WriteLine(BitConverter.ToString(span));
```

**Comment by GrabYourPitchforks** [comment]
FYI we don't normally add _startIndex_ and _length_ parameters on APIs which take spans. The caller is expected to slice the span appropriately before calling the method.

Do you have a usage example where you're relying on `BitConverter.ToString` in your own code? The example you provided won't compile, even with the APIs you had proposed here.

**Comment by Symbai** [comment]
Yes I do have but I cannot show it public. However I can show you a variation of it:

```C#
const int infoLength = 32;
byte[] infoStartingBytes = { 0xFF, 0xFF, 0xFF, 0xFF };

ReadOnlySpan<byte> span = System.IO.File.ReadAllBytes("C:\\data.dat").AsSpan();
int pos = span.IndexOf(infoStartingBytes);

while (pos != -1) {
	span = span.Slice(span.IndexOf(infoStartingBytes));
	int id = BitConverter.ToInt32(span);
	span = span.Slice(infoLength + 1);
	string name = System.Text.Encoding.UTF8.GetString(span);
	span = span.Slice(infoLength + 1);
	// .....
	string hash = BitConverter.ToString(span.Slice(infoLength).ToArray());
	span = span.Slice(infoLength + 1);

	pos = span.IndexOf(infoStartingBytes);
}
```

Using a span has the benefit of slicing. However at one point I have to convert it back to an array when I want to use BitConverter.ToString. I wouldn't mind if there is no overload with startIndex and length parameters since this can easily applied to the span instead but I want to get rid of the additional extra allocation of ToArray(). And I asked myself why every other function of BitConverter has received a span overload but not ToString?

**Comment by GrabYourPitchforks** [comment]
Thanks! I'll mark this ready for review.

**Comment by stephentoub** [comment]
Having the dashes between each byte is important to your use case?

**Comment by stephentoub** [comment]
Just want to point out that the details of the format used, dashes or not, casing, etc. was debated at length in https://github.com/dotnet/runtime/issues/17837, and the decision then was to just add Convert.ToHexString that only offered the one approach of no separators, all upper-case.

**Comment by Symbai** [comment]
> Having the dashes between each byte is important to your use case?

In my particular case it's important but from what I've seen the majority rather uses something like `BitConverter.ToString(bytes).Replace("-", "")`

But I think that when adding an overload, we shouldn't change the behavior. It should return exactly what the byte array overload returns. For me it makes more sense adding this overload with dashes and provide a separate method like Convert.ToHexString without dashes. It feels like these are use-cases for both, and this one here is just an overload for an existing function.

**Comment by stephentoub** [comment]
I'm not suggesting changing behavior.  Convert.ToHexString already takes a span and produces a hex string, just without dashes.

**Comment by XArtur0** [comment]
> I'm not suggesting changing behavior. Convert.ToHexString already takes a span and produces a hex string, just without dashes. I was questioning whether we need another.

So, all of us that do require the dash* have to either change our logic (not just code), roll our own hex formatter or keep allocating (and copying) memory on the heap? Don't get me wrong, all 3 are fair, its not like our existing code got broken...
I agree that #17837 is a better forward solution, I will probably end up changing my code to use it instead myself.

If no overload is added, perhaps `BitConverter.ToString` should be decorated with `[Obsolete("Use Convert.ToHexString where possible")]` so developers that remove the '-' easily realize they don't have to anymore =].

*I personally kept the dash because didn't want the performance hit of removing them.

**Comment by msschl** [comment]
> I'm not suggesting changing behavior. Convert.ToHexString already takes a span and produces a hex string, just without dashes. I was questioning whether we need another.

Well in my use case, I use the `BitConverter.ToString` method to print the hex string to the console for debug purposes. Having dashes keeps the string more readable for me. I often have to write some thing along this just for debugging:
```cs
Console.WriteLine(BitConverter.ToString(reader.ReadBytes(reader.BytesRemaining).ToArray()));
```
where `reader.ReadBytes(numOfBytes)` returns a `ReadOnlySpan<byte>`

I personally would appreciate adding this api.

**Comment by tannergooding** [comment]
Closing this as not planned.

As per the above, this was already debated extensively in a past API review and we opted to not go this route. There's not enough interest in this specific API and the overwhelming majority are using the APIs on `Convert` instead which don't involve any separator. Users who really need this and who cannot just use the existing `byte[] value, int startIndex, int length` overload can always copy the 20-30 lines of code required.
